CIP-hardforkSafeguard/README.md

```diff
@@ -0,0 +1,193 @@
+---
+CIP: ?
+Title: Hardfork Safeguard via Stake Representation
+Authors: Jared Corduan <jared.corduan@iohk.io>
+Status: Draft
+Type: Standards Track
+Created: 2022-06-30
+License: CC-BY-4.0
+---
+
+## Simple Summary / Abstract
+
+This CIP replaces a manual safety check regarding the readiness of the network
+for a hardfork with an automatic check.
+
+Ever since the Shelley ledger era, block headers have included a protocol
+version indicating the maximum supported protocol version that the block
+producer is capable of supporting (see section 13, Software Updates, of the
+[Shelley ledger specification](https://hydra.iohk.io/job/Cardano/cardano-ledger/shelleyLedgerSpec/latest/download-by-type/doc-pdf/ledger-spec)).
+This provides a helpful metric for determining how many blocks will be produced
+after a hardfork, since nodes that have not upgraded will no longer produce blocks.
+(Nodes that have not upgraded will fail the `chainChecks` check from Figure 74
+of the Shelley ledger specification, since the major protocol version in the
+ledger state will exceed the node's max major protocol version value,
+and hence can no longer make blocks.)
+
+If most of the blocks in the recent past (e.g. the last twenty-four hours) are
+broadcasting their readiness for a hardfork,
+we know that it is safe to propose an update to the major protocol version
+which triggers a hardfork.
+
+We propose automating this process.
+The ledger state will determine the stake
+(represented as the proportion of the active stake) of all the block producers
+whose last block contained the next major protocol version.
+Moreover, a new protocol parameter `hfThreshold` will be used to reject any
+protocol parameter update that proposes to change the major protocol version
+but does not have enough backing stake.
+
+## Motivation / History
+
+Currently, the governance key holders collectively agree to increase the major
+protocol version.  This allows them to make a human judgment as to the
+readiness of the network for a hardfork (using the mechanism described above).
+Since only a few, well-aligned parties are involved, this is currently easy to
+coordinate.  As we move into the Voltaire phase, where the governance of the
+network is decentralized,
+it is imperative that we codify this human judgment in the protocol itself.
+
+## Specification
+
+### New Protocol Parameter
+
+There will be a new protocol parameter named `hfThreshhold`,
```

Comment:
Note that making `hfThreshhold` a protocol parameter means that the quorum required for a change to protocol parameters has to be equal to or greater than the requirement to trigger a hardfork
Otherwise, you could first make a protocol parameter vote to lower `hfThreshhold` and then do the hardfork afterwards

Reply:
I completely get and empathize with this concern, but it does come at the cost of extra complexity (to an otherwise pretty bare-bones change). I'm not sure I understand what "has to be equal to or greater than the requirement" means (the quorum and the threshold have different units, they are not comparable), but I can guess that you would like the SPOs to be able to signal support for changing the `hfThreshhold` (the way the protocol version is done today). This would require a change to the block header, and we'd need to think about how it would be configured. The broadcast protocol version is baked into the node, and requires a software update to change (which is natural, given that it relates to a fork). Would we require a software upgrade to endorse a change to the `hfThreshhold`, or make it configurable?

---

Without addressing this concern, it makes `hfThreshhold` more like all the other protocol parameters that also greatly effect the SPOs, like `k`, the fees, etc. The governance structure has to openly, on chain for everyone to see, declare that a change is being made. And it is the SPOs who ultimately run the network. Decentralized governance is a huge topic that we'll be getting into next, but in the short term we can think about where to draw the line on small changes that we could implement sooner rather than later. My bias is toward being as minimal as possible, but I'm happy for folks to disagree!

Reply:
Intuitively, `hfThreshold` has a different feel than any other parameter. To me, it resembles an article of a constitution. Because of that, it would require a larger majority to get modified.

Without an protective mechanism (`hfThresold2`?), its value could go down over time, and there is surely a point where it starts getting dangerously low.